lib/types.js

```diff
@@ -1,10 +1,249 @@
 'use strict';
+const { formatLength, omit } = require('./util.js');
+
+const checkLength = (value, path, length) => {
+  const len = value.length;
+  const { min, max } = length;
+  if (min && len < min) return [`Field "${path}" value is too short`];
+  if (max && len > max) {
+    return [`Field "${path}" exceeds the maximum length`];
+  }
+};
+
+const scalar = {
+  kind: 'scalar',
+  check(value, path, def) {
+    if (typeof value !== this.name) {
+      return [`Field "${path}" is not of expected type: ${this.name}`];
+    }
+    const { length } = def;
+    if (length && this.length) {
+      const err = checkLength(value, path, length);
+      if (err) return err;
+    }
+    return [];
+  },
+  toLong(def) {
+    const { length } = def;
+    if (length) def.length = formatLength(length);
+    return def;
+  },
+};
+
+const enumerable = {
+  name: 'enum',
+  kind: 'scalar',
+  check(value, path, def) {
+    if (def.enum.includes(value)) return [];
+    return [`Field "${path}" value is not of enum: ${def.enum.join(', ')}`];
+  },
+  toLong(def) {
+    return { type: 'enum', ...def };
+  },
+};
+
+const reference = {
```

Comment:
IMO it would be good to separate the per-file and collect in some types/index.js